src/block/stdio_executor.rs

```diff
@@ -0,0 +1,453 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Copyright © 2019 Intel Corporation
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! A virtio block request execution abstraction.
+//!
+//! This module provides the following minimal abstraction for executing a virtio block device
+//! request:
+//!
+//! - [`StdIoBackend`](struct.StdIoBackend.html) which handles the execution of the block device
+//! request via [`StdIoBackend::execute`](struct.StdIoBackend.html#method.execute) method.
+//! The `StdIoBackend` is wrapping the block device backend and keeps the number of sectors of the
+//! backing file and its negotiated features too. This backend has to be, at least for now,
+//! `io::Read` and `io::Write`. In the future, we might add some abstraction for the file access
+//! operations.
+//!
+//! For more complex executors, that need asynchronous dispatch of requests for example, we can
+//! add separate modules for those abstractions as well.
+
+use std::fmt::{self, Display};
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::{io, result};
+
+use vm_memory::{Bytes, GuestMemory, GuestMemoryError};
+use vmm_sys_util::file_traits::FileSync;
+use vmm_sys_util::write_zeroes::{PunchHole, WriteZeroes};
+
+use crate::block::request::VIRTIO_BLK_T_FLUSH;
+use crate::block::request::{Request, RequestType};
+
+/// Read-only device.
+pub const VIRTIO_BLK_F_RO: u64 = 5;
+/// Flush command supported.
+pub const VIRTIO_BLK_F_FLUSH: u64 = 9;
+
+const SECTOR_SHIFT: u8 = 9;
+const SECTOR_SIZE: u64 = (0x01 as u64) << SECTOR_SHIFT;
+
+/// Trait that keeps as supertraits the ones that are necessary for the `StdIoBackend` abstraction
+/// used for the virtio block request execution.
+pub trait Backend: Read + Write + Seek + FileSync + PunchHole + WriteZeroes {}
+
+impl<B: Read + Write + Seek + FileSync + PunchHole + WriteZeroes> Backend for B {}
+
+/// Errors encountered during request execution.
+#[derive(Debug)]
+pub enum Error {
```

Comment:
Can we add something like [this](https://github.com/firecracker-microvm/firecracker/blob/c1ec34482f335aacee99901de472b5b0e53468d9/src/devices/src/virtio/block/request.rs#L32) method to ease converting an error to the appropriate status value?

Reply:
I wanted to give this a bit more thought in the next PR when I'll have a better idea of how to write the status too and regarding the method you've mentioned, for example, I would've liked to have some sort of `From` impl.
Anyway, I can quickly add the `status` method as is if it is really required for your PR in vmm-reference (but I expect to change it in the next PR).

Reply:
It's not immediately required, the temporary solution we're using is fine for the time being.